src/ripple/overlay/impl/MessageScheduler.cpp

```diff
@@ -0,0 +1,416 @@
+                {
+                    break;
+                }
+                auto channel = std::move(it->second->channel);
+                try
+                {
+                    it->second->receiver->onFailure(
+                        it->second->id, FailureCode::TIMEOUT);
+                }
+                catch (std::exception const&)
+                {
+                }
+                requests_.erase(it);
+                channels_.emplace_back(channel);
+            } while (false);
+        }
+        if (!channels_.empty() && !senders_.empty())
+        {
+            _offer(channels_, senders_);
+        }
+    };
+    {
+        std::lock_guard<std::mutex> lock(requestsMutex_);
+        request->timer.expires_after(timeout);
+        request->timer.async_wait(std::move(onTimer));
+        requests_.emplace(requestId, std::move(request));
+    }
+    // REVIEWER: Should we send before or after bookkeeping?
+    peer->send(packet);
+    JLOG(journal_.trace()) << "send,id=" << requestId << ",peer=" << peer->id();
+}
+
+void
+MessageScheduler::receive(std::shared_ptr<protocol::TMLedgerData> message)
+{
+    if (!message->has_requestcookie())
+    {
+        JLOG(journal_.warn()) << "LedgerData message missing request ID";
+        return;
+    }
+    auto requestId = message->requestcookie();
+    return _receive(requestId, message);
+}
+
+void
+MessageScheduler::receive(std::shared_ptr<protocol::TMGetObjectByHash> message)
+{
+    JLOG(journal_.trace()) << "receive,type=get_objects,count="
+                           << message->objects().size();
+    if (!message->has_seq())
+    {
+        JLOG(journal_.warn()) << "LedgerData message missing request ID";
+        return;
+    }
+    auto requestId = message->seq();
+    return _receive(requestId, message);
+}
+
+void
+MessageScheduler::_receive(
+    RequestId requestId,
+    std::shared_ptr<::google::protobuf::Message> message)
+{
+    push_value top(caller, "receive");
+    std::vector<Sender*> senders;
+    std::vector<Channel> channels;
+    {
+        // We push `tsenders` here to keep `onSuccess` callbacks from
+        // trying to lock offers after requests have been locked.
+        push_value top(tsenders, &senders);
+        std::lock_guard<std::mutex> lock(requestsMutex_);
+        auto it = requests_.find(requestId);
+        if (it == requests_.end())
+        {
+            // Either we never requested this data,
+            // or it took too long to arrive.
+            JLOG(journal_.warn()) << "unknown request ID: " << requestId;
+            return;
+        }
+        assert(it->second->id == requestId);
+        if (it->second->timer.cancel() < 1)
+        {
+            // Timer has already expired,
+            // and the `onFailure` callback has or will be executed.
+            return;
+        }
+        // TODO: Document assumption that callbacks are trivial.
+        // Non-trivial callbacks should just schedule a job.
```

Comment:
`onSuccess()` has document now. Please add this.

Reply:
It's fair to push back.  I suspect it would call for a significant redesign.

>  I feel it is essential to the design and usefulness of `MessageScheduler`.

I'm guessing you feel that way because you believe that the locking/unlocking overhead would be too significant?  Or is there a different motivation?  It would be good for me to understand the motivation behind the current design.  I could certainly be missing something.

> We deal all the time with components that place contracts on their use. Do not deference an iterator after it is invalidated. Do not double-free a pointer. Do not double-join a thread.

Truly.  But what you are describing are commonly known, and taught, constraints.  And there are specific coding patterns that people employ in order to help others avoid these contract violations.

- In C++ do not pass a pointer expecting that the called function will call `delete()` on that pointer.  Instead, use `unique_ptr`, `shared_ptr`, or `weak_ptr` to manage lifetime.
- Never pass an invalidated iterator to a function you're calling.
- For heaven's sake, don't `join()` a thread that you are not in charge of!

And usually these kinds of contracts are presented at the _input_ to a function:

- Don't _call_ `delete()` using a pointer whose memory you don't own.
- Don't _call_ any function passing an invalidated iterator.
- Don't _call_ `join()` on a thread you are not responsible for.

The contract you're describing is fundamentally different.  Here the _caller_ is imposing a constraint on application code that it is calling.  The first time that application code is coded, yes the developer has full visibility of the constraints and will probably follow them.

However once that code is in the code base it needs to be maintained.  The maintainer will frequently have no idea all of the various places that _might_ be calling a function.  Or whether one of those callers might impose special constraints.

There are indeed coding situations where callers apply the kinds of constraints on the callee that we're talking about.  Signal handlers leap to mind.  asio callbacks are another example.

Both signal handlers and asio callbacks are notorious for being hard to get right, hard to test, hard correctly maintain, and easy to accidentally break.

What I'm suggesting is that we shouldn't be adding to the number of places where people need to be _that_ careful in the rippled codebase.  

`MessageScheduler` is intended to (eventually) be a general facility that will be widely used for a large portion of the messaging needs in the server.  Let's not make it harder than necessary to correctly maintain the code that uses the facility.

Counterpoint?

Reply:
> I'm guessing you feel that way because you believe that the locking/unlocking overhead would be too significant?

Almost. It's not the overhead of unlocking the mutex that I'm worried about. I'm worried about the overhead of bookmarking the current state of the scheduler, transitioning to a state that is safe to unlock, then restoring the bookmarked state after re-locking. I expect this work to be unnecessary in the vast majority of cases, and I expect the implications of the scheduler being unable to assume it holds a lock until it finishes this work to add significant complexity to the design.

Separately, this is a fork point in the algorithm. Even if I were to safely unlock to call each callback, callbacks that do not schedule jobs are limiting the parallelism. I don't want to require _every_ callback to schedule a job, though. Callbacks that can finish more quickly than it takes to schedule a job should be able to.

> The contract you're describing is fundamentally different.

It could be framed this way: don't _call_ `schedule` or `send` with a callback that is non-trivial.

> There are indeed coding situations where callers apply the kinds of constraints on the callee that we're talking about. `asio` callbacks are another example.

These are callbacks, and in fact, they are indirectly `asio` callbacks. The complexity we're talking about is inherited.

> What I'm suggesting is that we shouldn't be adding to the number of places where people need to be that careful in the rippled codebase.

I don't like this restriction. I can agree it should be kept to a minimum, and well-documented, made easy to get right and hard to get wrong, but I don't like the idea of a total ban.

If the team _really_ wants to block this code on this point, then the alternative I would prefer is to have the scheduler schedule jobs for the callbacks. Then, those calls are guaranteed to be trivial and parallelized.

Reply:
See [additional comment](https://github.com/XRPLF/rippled/pull/4217#discussion_r1134488074).